src/groups/mqb/mqbauthn/mqbauthn_authenticationcontroller.cpp

```diff
@@ -0,0 +1,483 @@
+        rc_VALIDATE_ANONYMOUS_FAILED    = -3,
+        rc_DEFAULT_AUTHENTICATOR_FAILED = -4
+    };
+
+    const mqbcfg::AuthenticatorConfig& authenticatorConfig =
+        mqbcfg::BrokerConfig::get().authentication();
+
+    // Step 1: Collect all available plugin factories (built-in + external)
+    PluginFactories pluginFactories(d_allocator_p);
+    if (int rc = collectAvailablePluginFactories(&pluginFactories)) {
+        return rc * 10 + rc_COLLECT_FACTORIES_FAILED;  // RETURN
+    }
+
+    // Step 2: Create authenticators based on configuration
+    if (int rc = createConfiguredAuthenticators(errorDescription,
+                                                pluginFactories,
+                                                authenticatorConfig)) {
+        return rc * 10 + rc_CREATE_AUTHENTICATORS_FAILED;  // RETURN
+    }
+
+    // Step 3: Validate anonymous credential matches a configured authenticator
+    if (int rc = validateAnonymousCredential(errorDescription,
+                                             authenticatorConfig)) {
+        return rc * 10 + rc_VALIDATE_ANONYMOUS_FAILED;  // RETURN
+    }
+
+    // Step 4: Ensure at least one authenticator is available
+    if (int rc = ensureDefaultAuthenticator(errorDescription)) {
+        return rc * 10 + rc_DEFAULT_AUTHENTICATOR_FAILED;  // RETURN
+    }
+
+    BALL_LOG_INFO << "Successfully initialized " << d_authenticators.size()
+                  << " authenticator(s)";
+    return rc_SUCCESS;
+}
+
+int AuthenticationController::collectAvailablePluginFactories(
+    PluginFactories* pluginFactories)
+{
+    // Collect external plugin factories loaded by PluginManager
+    d_pluginManager_p->get(mqbplug::PluginType::e_AUTHENTICATOR,
+                           pluginFactories);
+
+    // Add built-in plugin factories
+    bsl::vector<mqbplug::PluginInfo>::const_iterator it =
+        d_builtInPluginLibrary->plugins().cbegin();
+    for (; it != d_builtInPluginLibrary->plugins().cend(); ++it) {
+        if (it->type() == mqbplug::PluginType::e_AUTHENTICATOR) {
+            pluginFactories->insert(it->factory().get());
+        }
+    }
+
+    return 0;
+}
+
+int AuthenticationController::createConfiguredAuthenticators(
+    bsl::ostream&                      errorDescription,
+    const PluginFactories&             pluginFactories,
+    const mqbcfg::AuthenticatorConfig& authenticatorConfig)
+{
+    enum RcEnum {
+        rc_SUCCESS             = 0,
+        rc_DUPLICATE_MECHANISM = -1,
+        rc_START_AUTHENTICATOR = -2,
+        rc_PLUGIN_NOT_FOUND    = -3
+    };
+
+    bmqu::MemOutStream errorStream(d_allocator_p);
+
+    // If authenticators are explicitly configured, only create those
+    const bsl::vector<mqbcfg::AuthenticatorPluginConfig>& authenticators =
+        authenticatorConfig.authenticators();
+    if (!authenticators.empty()) {
+        BALL_LOG_INFO << "Creating " << authenticators.size()
+                      << " explicitly configured authenticator(s)";
+
```

Comment:
Good to use an early return trick here to reduce indentation level of a big chunk of code following


```suggestion
    if (authenticators.empty()) {
        BALL_LOG_INFO << "No configured authenticator(s)";
        return rc_SUCCESS;
    }

    BALL_LOG_INFO << "Creating " << authenticators.size()
                  << " explicitly configured authenticator(s)";
```